File: files/unminified/static/js/components/catalog.js
export const name = 'catalog';

import * as Util from "/static/js/util.js"
import createEntries from "/static/js/components/entries.js"

export default function (entries, progress) {
    console.debug("catalog entries:", entries)
    console.debug("catalog progress:", progress)

    let urlFunc = (i) => {
        return `/api/series/${entries[i].sid}/cover?thumbnail=true`
    }
    let idFunc = (i) => {
        return entries[i].sid
    }
    let extra = {
        fmtProgress(e) {
            let prog = this.progress[e.sid]
            let pages = e.num_pages
            return `Progress: ${Util.Fmt.SeriesPercent(prog, pages)}`
        }
    }

    return createEntries(entries, progress, urlFunc, idFunc, extra)
}

File: pkg/progress/catalog.go
package progress

import "encoding/json"

type Catalog struct {
	m map[string]Series
}

func NewCatalogProgress() *Catalog {
	return &Catalog{m: map[string]Series{}}
}

func (cp *Catalog) Add(sid string, p Series) {
	cp.m[sid] = p
}

func (cp Catalog) MarshalJSON() ([]byte, error) {
	return json.Marshal(cp.m)
}
